app/Http/Controllers/DealController.php

```
@@ -1622,17 +1592,20 @@ public function update(UpdateRequest $request, $id)
-        $newProductIds = array_diff($request->product_id ?? [], $oldProductIds);
-        $removedProductIds = array_diff($oldProductIds, $request->product_id ?? []);
+        $oldProductIds = $deal->products()->pluck('products.id')->toArray();
 
         // Same as packages above: only sync products when the field is present.
+        // The diff below must be scoped the same way — without it, a request
+        // that omits product_id entirely reads as "every existing product was
+        // removed" and fires notifyPropertyUnlinked for all of them.
+        $newProductIds = [];
+        $removedProductIds = [];
         if ($request->has('product_id')) {
             $deal->products()->sync(is_array($request->product_id) ? $request->product_id : []);
```

PHPStan (2.2.8)

[error] Cannot call method products() on array<Illuminate\Database\Eloquent\Builder>|Illuminate\Database\Eloquent\Builder|Illuminate\Database\Eloquent\Collection|Illuminate\Database\Eloquent\Model.


(method.nonObject)